versions/en/toml-v0.4.x.abnf

```diff
@@ -0,0 +1,321 @@
+MultilineLiteralString
+  = 3SingleQuote
+    [ Newline ]  ; This optional newline is not added to the string.
+    *( MultilineLiteralCharacter
+     / Newline
+     )
+    3SingleQuote
+
+; Control-characters except HTAB, are not allowed.
+MultilineLiteralCharacter
+  = %x09  ; Horizontal tab "	"
+  / %x20-10FFFF
+
+Boolean
+  = True
+  / False
+
+True
+  = %x74.72.75.65     ; "true"
+
+False
+  = %x66.61.6C.73.65  ; "false"
+
+; A 64-bit signed integer.
+Integer
+  = [ Plus / Minus ] IntegerDigits
+
+Plus
+  = %x2B  ; Plus sign +
+
+Minus
+  = %x2D  ; Minus sign -
+
+IntegerDigits
+  = Digit
+  / Digit1to9  ; Decimal numbers except 0 should not start with "0".
+    1*( [ Underscore ] Digit )  ; Use "_" as delimiters.
+
+ZeroPrefixableInteger
+  = Digit
+    *( [ Underscore ] Digit )
+
+; A double-precision 64-bit floating-point number in IEEE 754 standard.
+Float
+  = Integer
+    ( Fraction [ Exponent ]
+    / Exponent
+    )
+
+Fraction
+  = DecimalPoint ZeroPrefixableInteger
+
+DecimalPoint
+  = %x2E  ; .
+
+Exponent
+  = E Integer
+
+E
+  = %x65  ; e
+  / %x45  ; E
+
+; It should conform to RFC 3339? Really?
+; RFC 3339: http://www.ietf.org/rfc/rfc3339.txt
+DateTime
+  = FullDate T FullTime
+
+T
+  = %x54  ; T, stands for "time"
+  ; The lower case "t" is not allowed?
+
+FullDate
+  = Year "-" Month "-" MDay
+
+Year
+  = 4Digit  ; Unix time? 1970-????
+
+Month
+  = 2Digit  ; 01-12
+
+MDay
+  = 2Digit  ; 01-31, based on the year and the month
+
+FullTime
+  = Time TimeOffset
+
+Time
+  = Hour ":" Minute ":" Second [ SecondFraction ]
+
+Hour
+  = 2Digit  ; 00-23
+
+Minute
+  = 2Digit  ; 00-59
+
+Second
+  = 2Digit  ; 00-59, even 00-58 or 00-60, based on leap second rules?
+
+SecondFraction
+  = "." 1*Digit
+
+TimeOffset
+  = Z
+  / ( "+" / "-" ) Hour ":" Minute
+
+Z
+  = %x5A  ; Z, stands for "UTC time"
+  ; The lower case "z" is not allowed?
+
+Array
+  = LeftBracket *ArraySpace
+    [ ArrayValue *ArraySpace
+      [ Comma *ArraySpace ]  ; Allow one extra comma before the right bracket.
+    ]
+    RightBracket
+
+LeftBracket
+  = %x5B  ; [
+
+RightBracket
+  = %x5D  ; ]
+
+ArraySpace
+  = Whitespace
+  / Newline
+  / Comment
+
+; Values in the same array should be of the same type.
+; For simplicity, I don't further specify it here. Sorry.
+ArrayValue
+  = Value
+    [ *ArraySpace Comma *ArraySpace ArrayValue ]
+
+InlineTable
+  = LeftBrace *Whitespace
+    [ KeyValue
+      *( *Whitespace Comma *Whitespace KeyValue )
+      *Whitespace
+    ]
+    RightBrace
+
+LeftBrace
+  = %x7B  ; {
+
+RightBrace
+  = %x7D  ; }
+
+TableHeader
+  = LeftBracket *Whitespace
+    Key *( *Whitespace "." *Whitespace Key )
+    *Whitespace RightBracket
+
+TableArrayHeader
+  = LeftBracket TableHeader RightBracket
+
+DoubleQuote
+  = %x22  ; "
+
+SingleQuote
+  = %x27  ; '
+
+Backslash
+  = %x5C  ; \
+
+Comma
+  = %x2C  ; ,
+
+Underscore
+  = %x5F  ; _
+
+Letter
+  = %x41-5A  ; "A" to "Z"
```

Comment:
Well, you'd better ask the original designers of TOML 0.4.0, because I just wanted to get more details of 0.4.0 depicted.
PS. The chart of Unicode characters is just a mess. Be careful. Also, my grammar still have bugs, but I've no interest to improve it any more.